adminSiteServer/apiRoutes/dataInsights.ts

```diff
@@ -0,0 +1,294 @@
+import * as db from "../../db/db.js"
+import { getTagsGroupedByGdocId } from "../../db/model/Gdoc/GdocFactory.js"
+import { mapQueryParamToChartTypeName } from "@ourworldindata/grapher"
+import { getTimeDomainFromQueryString } from "@ourworldindata/utils"
+
+const GRAPHER_URL_PREFIX = "https://ourworldindata.org/grapher/"
+const EXPLORER_URL_PREFIX = "https://ourworldindata.org/explorers/"
+
+type DataInsightRow = Pick<
+    DbRawPostGdoc,
+    "slug" | "content" | "published" | "publishedAt" | "markdown"
+> &
+    Pick<DbRawImage, "cloudflareId" | "filename" | "originalWidth"> & {
+        gdocId: DbRawPostGdoc["id"]
+        narrativeChartId?: DbPlainChartView["id"]
+        narrativeChartConfigId?: DbPlainChartView["chartConfigId"]
+        chartConfig?: DbRawChartConfig["full"]
+        imageId?: DbRawImage["id"]
+        tags?: MinimalTag[]
+    }
+
+export async function getAllDataInsightIndexItems(
+    req: Request,
+    res: e.Response<any, Record<string, any>>,
+    trx: db.KnexReadonlyTransaction
+) {
+    return getAllDataInsightIndexItemsOrderedByUpdatedAt(trx)
+}
+
+async function getAllDataInsightIndexItemsOrderedByUpdatedAt(
+    knex: db.KnexReadonlyTransaction
+): Promise<OwidGdocDataInsightIndexItem[]> {
+    const rows = await db.knexRaw<DataInsightRow>(
+        knex,
+        `-- sql
+
+        -- only consider published stand-alone charts since we join by slug which is only unique for published charts
+        WITH published_charts AS (
+            SELECT cc.id, cc.slug, cc.full
+            FROM chart_configs cc
+            JOIN charts c ON c.configId = cc.id
+            WHERE cc.full ->> '$.isPublished' = 'true'
+        )
+
+        SELECT
+            -- gdoc fields
+            pg.id AS gdocId,
+            pg.slug,
+            pg.content,
+            pg.published,
+            pg.publishedAt,
+            pg.markdown,
+
+            -- narrative chart fields
+            cw.id AS narrativeChartId,
+            cw.chartConfigId AS narrativeChartConfigId,
+
+            -- chart config (prefer narrative charts over grapher URLs)
+            COALESCE(cc_chartView.full, cc_grapherUrl.full) AS chartConfig,
+
+            -- image fields
+            i.id AS imageId,
+            i.filename,
+            i.cloudflareId,
+            i.originalWidth
+        FROM posts_gdocs pg
+
+        -- extract the slug from the given Grapher URL and join by it
+        LEFT JOIN published_charts cc_grapherUrl
+            ON cc_grapherUrl.slug = SUBSTRING_INDEX(SUBSTRING_INDEX(content ->> '$."grapher-url"', '/grapher/', -1), '\\?', 1)
+
+        -- join the chart_views table to get the config of the narrative chart
+        LEFT JOIN chart_views cw
+            ON cw.name = content ->> '$."narrative-chart"'
+        LEFT JOIN chart_configs cc_chartView
+            ON cc_chartView.id = cw.chartConfigId
+
+        -- join the images table by filename (only works for data insights where the image block comes first)
```

Comment:
I don't think any author has ever _not_ done this, but this will be the first time we explicitly hinge business logic on this assumption - so maybe we should now add validation to assert this?

Reply:
Yeah, that's a good idea! I was debating whether it's okay to make this assumption here. I finally settled on this because writing more code to cover cases outside our guidelines didn't make much sense to me. Good call to make it explicit!